### PR TITLE
feat: Add basic support for git tag path prefixes

### LINKED ISSUE #20949: Allow the use of path-prefixed sever tags as a valid target revision
# Summary

Many mono Repos (ex: [bitami charts](https://github.com/bitnami/charts/tags)) organize semver release tags by prefixing them with an identifier (ex: `component/v1.0.0`). Currently, this tag format cannot be parsed when used in a version constraint expression (ex: `targetRevision: ">= component/v1.0.0"`)

Creating tags is the standard way to release software on GitHub and because of the popularity of monorepos, either for software components or hosting helm charts, I believe Argo CD should be able to correctly parse this type of tag when used in a version constraint expression.

# Motivation

Developers that are using this tagging strategy for managing releases have no good alternatives:

- **Use a static tag**: releasing a new version requires `git push --tags --force` which requires special permissions and can inadvertently overwrite other tags.
- **Modify the target revision in the Application resource manifest**: Manifests are often hosted in a separate branch or repository, making the release process more complicated than simply creating a new tag.
- **Use a separate branch or repo for managing Argo application releases**: Again, more complicated than simply tagging a commit. Developers end up with a more complicated release process.

This change also provides opportunities for new deployment strategies:
- use tags to differentiate an app releases and a deployment. For example, only tags prefixed like this trigger a release: `argocd/v1.0.0`
- use tags to control specific deployments: `prod/v1.0.0`, `staging/v1.0.0`
- route deployments to specific clusters: `clusterA/v1.0.0`, `clusterB/v1.0.0`

# How do you think this should be implemented?
Use regex to extract path prefixes out of tags in constraint strings so they can be processed. Ensure prefixes are considered during matching.

### PR DESCRIPTION
**Pull request by chzar**
TL;DR This PR changes the resolveSemverRevision() function to tolerate a path prefix before the semantic version.

```
# example
targetRevision: '>=foo/bar/v0.0.0-rc0'
```

This change is useful for teams that:
- want to use hierarchical tags - useful for monorepos (app1/v0.0.0, app2/v1.0.0)
- want to use a hierarchical namespace to route argo deployments using generators (app1/cluster1/prod/v0.0.0, app1/cluster2/dev/v0.0.0)
- use git flow or other tag based workflows with path prefixes

Closes #20949

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by chzar** [comment]
To me this change is not so much a feature as a logic change to tolerate tag prefixes - therefor I have not created an enhancement proposal. Happy to do so if the maintainer believe otherwise.

**Comment by chzar** [comment]
I want to improve this PR to move the constraints to the left side of the targetRevision string and support multiple constraints. Standby...

**Review by andrii-korotkov-verkada — APPROVED** [review]

**Comment by andrii-korotkov-verkada** [comment]
Please, sign off your commits.

**Comment by chzar** [comment]
I signed my commits and added a mention of this functionality in the docs under tracking strategies: use cases.

**Comment by chzar** [comment]
@andrii-korotkov-verkada

Just checking if I need to do anything else or just sit tight and wait for a review?

**Comment by andrii-korotkov-verkada** [comment]
Just wait for review. You may show up at the contributors meeting and introduce your changes.

**Comment by dplacencia** [comment]
Hi, any progress on this new feature?

**Comment by chzar** [comment]
> Hi, any progress on this new feature?

@dplacencia This feature is complete, but the maintainers have asked for more context before giving the greenlight. You can help out by voicing your support in issue #20949

**Review by agaudreault — CHANGES_REQUESTED** [review]
Wrong PR

**Comment by k4r1** [comment]
Hey @chzar, this is great - will be really useful to get this in. Since there's been a refactor around semver in #23310, I've taken the liberty of updating your logic to work with the new structure. I've opened a new PR referencing this one to help get it across the line

**Comment by chzar** [comment]
> Hey @chzar, this is great - will be really useful to get this in. Since there's been a refactor around semver in #23310, I've taken the liberty of updating your logic to work with the new structure. I've opened a new PR referencing this one to help get it across the line

**Comment by chzar** [comment]
@k4r1 Thanks for the assist, if you need anything from me - just let me know.

**Comment by chzar** [comment]
closing in favor of #26207.
